File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        stack = []
        cur = root
        while(True):
            while(cur is not None):
                stack.append(cur)
                cur = cur.left
            k-=1
            min_val = stack.pop()
            if k==0:
                return min_val.val
            cur = min_val.right
        
        return -1

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        Stack<TreeNode> stk = new Stack<>();
        
        TreeNode cur = root;
        
        while(true){
            while(cur!=null){
                stk.push(cur);
                cur = cur.left;
            }
            TreeNode smallest = stk.pop();
            k-=1;
            if(k==0){
                return smallest.val;
            }
            cur = smallest.right;
        }
        
    }
}